src/honey_badger.rs

```diff
@@ -200,43 +222,213 @@ where
+    /// this epoch.
+    fn all_selections_decrypted(&mut self) -> bool {
+        let ciphertexts = match self.ciphertexts.entry(self.epoch) {
+            Entry::Occupied(entry) => entry.into_mut(),
+            Entry::Vacant(entry) => entry.insert(BTreeMap::new()),
+        };
+        let all_ciphertext_proposers: BTreeSet<_> = ciphertexts.keys().collect();
+        let all_decrypted_selection_proposers: BTreeSet<_> =
+            self.decrypted_selections.keys().collect();
+        all_ciphertext_proposers.is_subset(&all_decrypted_selection_proposers)
+    }
+
+    /// Tries to decrypt the selection of transactions from a given proposer. Outputs `true` if and
+    /// only of decryption finished without errors.
+    fn try_decrypt_proposer_selection(&mut self, proposer_id: &NodeUid) -> bool {
+        let shares = &self.received_shares[&self.epoch][proposer_id];
+        if shares.len() > self.netinfo.num_faulty() {
+            if let Some(ciphertexts) = self.ciphertexts.get(&self.epoch) {
+                if let Some(ciphertext) = ciphertexts.get(proposer_id) {
+                    let ids_u64: BTreeMap<&NodeUid, u64> = shares
+                        .keys()
+                        .map(|id| (id, *self.netinfo.node_index(id).unwrap() as u64))
+                        .collect();
+                    let indexed_shares: BTreeMap<&u64, _> = shares
+                        .into_iter()
+                        .map(|(id, share)| (&ids_u64[id], share))
+                        .collect();
+                    if let Ok(decrypted_selection) = self
+                        .netinfo
+                        .public_key_set()
+                        .decrypt(indexed_shares, ciphertext)
+                    {
+                        self.decrypted_selections
+                            .insert(proposer_id.clone(), decrypted_selection);
+                        return true;
+                    }
+                }
+            }
+        }
+        false
+    }
+
+    fn send_decryption_shares(
+        &mut self,
+        cs_output: BTreeMap<NodeUid, Vec<u8>>,
+    ) -> Result<(), Error> {
+        for (proposer_id, v) in cs_output {
+            let mut ciphertext: Ciphertext<Bls12>;
+            if let Ok(ct) = bincode::deserialize(&v) {
+                ciphertext = ct;
+            } else {
+                warn!("Invalid ciphertext from proposer {:?} ignored", proposer_id);
+                // TODO: Log the incorrect node `j`.
+                continue;
+            }
+            if let Some(share) = self.netinfo.secret_key().decrypt_share(&ciphertext) {
+                // Send the share to remote nodes.
+                self.messages.0.push_back(
+                    Target::All.message(
+                        MessageContent::DecryptionShare {
+                            proposer_id: proposer_id.clone(),
+                            share: share.clone(),
```

Comment:
Those shares can be large. I wonder whether we should actually put them into an `Rc` here instead of cloning them? (Probably rather some optimization to try out and benchmark later…)

Reply:
I've been experimenting and this does indeed seem possible. Please check how you like the extra trait bounds!

I have one extra `clone` though, here in `send_decryption_shares`, where I move the share to an `Rc` and clone the `Rc` in the messages, and then I have to use the same share to pass by value to `handle_decryption_share_message`.

Also I will have to fix the tests tomorrow.